File: examples/integrateODEs.c
/*
  Last changed Time-stamp: <2006-10-14 07:07:07 raim>
  $Id: integrateODEs.c,v 1.2 2006/10/14 05:10:27 raimc Exp $
*/
/* 
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2.1 of the License, or
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. The software and
 * documentation provided hereunder is on an "as is" basis, and the
 * authors have no obligations to provide maintenance, support,
 * updates, enhancements or modifications.  In no event shall the
 * authors be liable to any party for direct, indirect, special,
 * incidental or consequential damages, including lost profits, arising
 * out of the use of this software and its documentation, even if the
 * authors have been advised of the possibility of such damage.  See
 * the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *
 * The original code contained here was initially developed by:
 *
 *     Rainer Machne
 *
 * Contributor(s):
 */

#include <stdio.h>
#include <stdlib.h>

#include <sbmlsolver/odeSolver.h>


int
main (int argc, char *argv[])
{
  int neq, nass, nconst;
  char *x[3];
  double x0[3];
  char *ode[2];
  ASTNode_t *ast[2];
  
  double time, rtol, atol;
  double printstep;

  /* SBML model containing events */
  Model_t *events;
  
  /* SOSlib types */
  odeModel_t *om;
  cvodeSettings_t *set;
  integratorInstance_t *ii;
  variableIndex_t *vi;

  /* parsing input ODEs and settings */
  /* time and error settings */
  time = 0.5;
  rtol = atol = 1.0E-9;
  printstep = 10;
  /* variables */
  neq = 2;
  x[0] = "C_cy"; 
  x0[0] = 0.0;
  ode[0] = "((150.0 * (3.8 - (p * D_cy) - (p * C_cy)) * (1.0 - (p * C_cy))) - (9.0 * C_cy))";
  x[1] = "D_cy";
  x0[1] = 9.0;
  ode[1] = "(3.8 - (3.0 * D_cy) - (p * D_cy) - (p * C_cy))";
  /* parameters */
  nconst = 1;
  x[2] = "p";
  x0[2] = 0.2;
  /* assignments */
  nass = 0;
  /* SBML model containing events */
  events = NULL;
  
  /* creating ODE ASTs from strings */
  ast[0] = SBML_parseFormula(ode[0]);
  ast[1] = SBML_parseFormula(ode[1]);

  /* end of input parsing */
  
  /* Setting SBML ODE Solver integration parameters  */
  set = CvodeSettings_createWithTime(time, printstep);
  CvodeSettings_setErrors(set, atol, rtol, 1e4);
  CvodeSettings_setStoreResults(set, 0);

  /* activating default sensitivity */
  CvodeSettings_setSensitivity(set, 1);
  
  /* creating odeModel */
  /* `events' could be an SBML model containing events */
  om = ODEModel_createFromODEs(ast, neq, nass, nconst, x, x0, events);

  /* creating integrator */
  ii = IntegratorInstance_create(om,set);

  /* integrate */
  printf("integrating:\n");
  while( !IntegratorInstance_timeCourseCompleted(ii) )
  {
    if ( !IntegratorInstance_integrateOneStep(ii) )
    {
      SolverError_dump();
      break;
    }
    else
      IntegratorInstance_dumpData(ii);    
  }

  /* Specific data interfaces: */
  /* names are stored in the odeModel and values 
     are independently stored in the integrator: you can
     get different values from different integrator instances
     built from the same model */
  vi = ODEModel_getVariableIndex(om, "C_cy");
  printf("\nVariable %s has final value of %g at time %g\n\n",
	 ODEModel_getVariableName(om, vi),
	 IntegratorInstance_getVariableValue(ii, vi),
	 IntegratorInstance_getTime(ii));
  VariableIndex_free(vi);

  vi = ODEModel_getVariableIndex(om, "p");
  printf("Sensitivies to %s:\n",  ODEModel_getVariableName(om, vi));
  IntegratorInstance_dumpPSensitivities(ii, vi);
  printf("\n\n");
  VariableIndex_free(vi);

  /* finished, free stuff */
  /* free ASTs */
  ASTNode_free(ast[0]);
  ASTNode_free(ast[1]);
  /* free solver structures */
  IntegratorInstance_free(ii);
  ODEModel_free(om);  
  CvodeSettings_free(set);
  
  return (EXIT_SUCCESS);  
}



/* End of file */

File: tutorial/integrateODES.pl
#!/usr/bin/perl
# -*-Perl-*-
# Last changed Time-stamp: <2007-09-26 16:23:22 xtof>
# $Id: integrateODES.pl,v 1.1 2007/09/28 13:14:16 raimc Exp $

use ODES;
use LibSBML;
use strict;

#my $ode0 = 'a-(b+1)*u+v*u^2'; # u'
#my $ode1 = 'b*u-v*u^2';       # v'
#my @x  = qw/u v a b/;
#my @x0 = (1.0, 1.5, 1.0, 1.5);

my $ode0 = '((150.0 * (3.8 - (p * D_cy) - (p * C_cy)) * (1.0 - (p * C_cy))) - (9.0 * C_cy))';
my $ode1 = '(3.8 - (3.0 * D_cy) - (p * D_cy) - (p * C_cy))';

# use LibSBML function to convert the formula into
# an abstract syntax tree (AST) representation
my $astA = LibSBML::parseFormula($ode0);
my $astB = LibSBML::parseFormula($ode1);
my @AST = ($astA, $astB);
my $neq = scalar @AST;
my @x = qw/C_cy D_cy p/;
my @x0 = (0.0, 9.0, 0.2);

my $om = ODES::ODEModel_createFromODEs(\@AST, 2, 0, 1, \@x, \@x0, undef);

my $settings = ODES::CvodeSettings_create();
$settings->CvodeSettings_setTime(0.5, 10);
$settings->CvodeSettings_setSensitivity(1);

my $ii = ODES::IntegratorInstance_create($om, $settings);

while( ! $ii->IntegratorInstance_timeCourseCompleted() ) {
  if ( ! $ii->IntegratorInstance_integrateOneStep() ) {
    ODES::SolverError_dump();
    last;
  }
  else {
    $ii->IntegratorInstance_dumpData();
  }
}

my $vi = $om->ODEModel_getVariableIndex('C_cy');

printf "\nVariable %s has final value of %g at time %g\n\n",
         $om->ODEModel_getVariableName($vi),
         $ii->IntegratorInstance_getVariableValue($vi),
         $ii->IntegratorInstance_getTime();

$vi->VariableIndex_free();

$vi = $om->ODEModel_getVariableIndex('p');

printf "Sensitivies to %s:\n",
    $om->ODEModel_getVariableName($vi);

$ii->IntegratorInstance_dumpPSensitivities($vi);
print "\n";

$om->ODEModel_free();
$settings->CvodeSettings_free();
$vi->VariableIndex_free();

#---
sub substitute_params {
  my ($eq, $params) = @_;
  for my $key (keys %$params) {
    $eq =~ s/$key/$params->{$key}/g;
  }

  return $eq;
}

__END__
my $sbmlfile = '../../examples/MAPK.xml';

my $rd = new LibSBML::SBMLReader;
my $d  = $rd->readSBML($sbmlfile);

my $settings = ODES::CvodeSettings_create();
$settings->CvodeSettings_setTime(10000, 100);

my $results = ODES::SBML_odeSolver($d, $settings);

$results->SBMLResults_dump();

